test/Microsoft.AspNetCore.Mvc.Core.Test/Internal/ControllerActionDescriptorProviderTests.cs

```diff
@@ -1483,17 +1485,49 @@ public void GetDescriptors_SplitsConstraintsBasedOnControllerRoute_Override()
             var options = new TestOptionsManager<MvcOptions>();
             options.Value.Conventions.Add(convention);
 
-            var controllerTypeProvider = new StaticControllerTypeProvider(new[] { controllerTypeInfo });
+            var manager = GetApplicationManager(new[] { controllerTypeInfo });
+
             var modelProvider = new DefaultApplicationModelProvider(options);
 
             var provider = new ControllerActionDescriptorProvider(
-                controllerTypeProvider,
+                manager,
                 new[] { modelProvider },
                 options);
 
             return provider;
         }
 
+        private static ApplicationPartManager GetApplicationManager(IEnumerable<TypeInfo> controllerTypes)
+        {
+            var manager = new ApplicationPartManager();
+            manager.ApplicationParts.Add(new TestPart(controllerTypes));
+            manager.FeatureProviders.Add(new TestProvider());
+            return manager;
+        }
+
+        private class TestPart : ApplicationPart, IApplicationPartTypeProvider
+        {
+            public TestPart(IEnumerable<TypeInfo> types)
+            {
+                Types = types;
+            }
+
+            public override string Name => "Test";
+
+            public IEnumerable<TypeInfo> Types { get; }
+        }
+
+        private class TestProvider : IApplicationFeatureProvider<ControllerFeature>
```

Comment:
TestPart and TestProvider i've been placing them adhoc where I needed them for tests, where would this go?